pkg/controller/job/job_controller.go

```diff
@@ -421,6 +423,33 @@ func (jm *Controller) updateJob(old, cur interface{}) {
 	}
 }
 
+// deleteJob enqueues the job and all the pods associated with it that still
+// have a finalizer.
+func (jm *Controller) deleteJob(obj interface{}) {
+	jm.enqueueController(obj, true)
+	jobObj, ok := obj.(*batch.Job)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %+v", obj))
+			return
+		}
+		jobObj, ok = tombstone.Obj.(*batch.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a job %+v", obj))
+			return
+		}
+	}
+	// Listing pods shouldn't really fail, as we are just querying the informer cache.
+	pods, _ := jm.podStore.Pods(jobObj.Namespace).List(labels.Everything())
+	for _, pod := range pods {
+		controllerRef := metav1.GetControllerOf(pod)
+		if (controllerRef == nil || controllerRef.UID == jobObj.UID) && hasJobTrackingFinalizer(pod) {
+			jm.enqueueOrphanPod(pod)
```

Comment:
are we relying on observing the job deletion to handle orphan pods correctly? what if the controller is down when the job is deleted and the pods are orphaned. does it remove finalizers correctly if it starts up and observes its finalizer on pods which aren't owned by a job?

Reply:
That's the other fix in function `deletePod`. If it's not being deleted, it could still be adopted by other jobs.

I don't think I can enqueue the pod for finalizer removal in that scenario. This race condition could happen: `syncJob` for another job adopts it and adds a finalizer. `syncOrphanPod` runs and deletes the finalizer.

Reply:
I realized that I'm contradicting myself in the comment above with respect to the current implementation.

I need to think about this more.

/hold

Reply:
All good now
/hold cancel